prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,675 @@
+			}
+			goto found;
+		}
+
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"During sm2 allocation of space for endpoint named %s "
+			"an existing conflicting address was found at AV[%d]\n",
+			name, jentry);
+
+		if (!pid_lives(entries[jentry].pid)) {
+			FI_WARN(&sm2_prov, FI_LOG_AV,
+				"The pid which allocated the conflicting AV is dead.  "
+				"Reclaiming as our own.\n");
+			// it is possible that EP's referencing this region are
+			// still alive... don't know how to check (they likely
+			// died if PID died)
+			goto found;
+		}
+		else {
+			FI_WARN(&sm2_prov, FI_LOG_AV,
+				"ERROR: The endpoint (pid: %d) with conflicting address "
+				"%s is still alive.\n",
+				entries[jentry].pid, name);
+			return -FI_EADDRINUSE;
+		}
+	}
+
+	/* fine, we could not find the entry, so now look for an empty slot */
+	for (jentry = 0; jentry < header->ep_enumerations_max; jentry++) {
+		peer_pid = entries[jentry].pid;
+		if (peer_pid == 0)
+			goto found;
+		else if (peer_pid < 0) {
+			// A third peer might have entered this address into their AV,
+			// and there is no current way to check this... need to keep this
+			// entry in the file until we clean up
+			continue;
+		}
+		else {
+			if (!pid_lives(peer_pid)) {
+				sm2_coordinator_extend_for_entry(map, jentry);
+				entries = sm2_mmap_entries(map);
+				header = (void *)map->base;
+				struct sm2_region *peer_region =
+				    sm2_mmap_ep_region(map, jentry);
+
+				if (entries[jentry].startup_ready &&
+				    smr_freestack_isfull(sm2_free_stack(peer_region))) {
+					/* we found a slot with a dead PID and the
+					 * freestack is full */
+					entries[jentry].pid = 0;
+					goto found;
+				};
+			}
+		}
+	}
+
+	fprintf(stderr,
+		"No available entries were found in the coordination file, all %d were "
+		"used\n",
+		header->ep_enumerations_max);
+	return -FI_EAVAIL;
+
+found:
+
+	if (self) {
+		entries[jentry].startup_ready = 0;
+		atomic_wmb();
+		entries[jentry].pid = pid;
+	}
+
+	if (!self && entries[jentry].pid == 0) {
+		entries[jentry].startup_ready = 0;
+		atomic_wmb();
+		entries[jentry].pid = -pid;
+	}
+
+	FI_WARN(&sm2_prov, FI_LOG_AV, "Allocated sm2 region for %s at AV[%d]\n", name,
+		jentry);
+	strncpy(entries[jentry].ep_name, name, SM2_NAME_MAX);
+
+	*av_key = jentry;
+
+	/* With the entry allocated, we now need to ensure it's mapped. */
+	if (!sm2_mapping_long_enough_check(map, jentry)) {
+		sm2_coordinator_extend_for_entry(map, jentry);
+	}
+	return 0;
+}
+
+/**
+ * @brief look-up the name.
+ *
+ * @param[in] name
+ * @param[in] map
+ * @return the index of the name, or -1
+ */
+int sm2_coordinator_lookup_entry(const char *name, struct sm2_mmap *map)
+{
+
+	struct sm2_coord_file_header *header = (void *)map->base;
+	struct sm2_ep_allocation_entry *entries;
+	int jentry;
+
+	entries = sm2_mmap_entries(map);
+
+	for (jentry = 0; jentry < header->ep_enumerations_max; jentry++) {
+		if (0 == strncmp(name, entries[jentry].ep_name, SM2_NAME_MAX)) {
```

Comment:
Doing a O(n) walk through to see if the name matches is not very efficient. I recommend replacing this lookup with a rbmap like the current smr_map uses.

Reply:
This is startup, and not used on the critical path.

Reply:
On the critical path, we have:
```
sm2_mmap_remap() // on very rare occasions after late arrivers join
```